otter/util/zkpartitioner.py

```diff
@@ -0,0 +1,133 @@
+"""
+ZooKeeper set-partitioning stuff.
+"""
+
+from twisted.application.internet import TimerService
+from twisted.application.service import MultiService
+from twisted.internet.defer import succeed
+
+
+class Partitioner(MultiService, object):
+    """
+    A Twisted service which uses a Kazoo :obj:`SetPartitioner` to allocate
+    logical ``buckets`` between nodes.
+
+    Multiple servers can instantiate these with the same ``partitioner_path``
+    and ``buckets``, and each server will get a disjoint subset of
+    ``range(buckets)`` allocated to them.
+
+    In order to be notified of which buckets are allocated to the current node,
+    a ``got_buckets`` function must be passed in which will be called when the
+    local buckets have been determined.
+    """
+    def __init__(self, kz_client, interval, partitioner_path, buckets,
+                 time_boundary, log, got_buckets,
+                 clock=None):
+        """
+        :param log: a bound log
+        :param kz_client: txKazoo client
+        :param partitioner_path: ZooKeeper path, used for partitioning
+        :param list buckets: buckets to distribute between nodes. Ideally
+            there should be at least as many elements as nodes taking part in
+            this partitioner. This should be a sequence of str.
+        :param time_boundary: time to wait for partitioning to stabilize.
+        :param got_buckets: Callable which will be called with a list of
+            buckets when buckets have been allocated to this node.
+        :param clock: clock to use for checking the buckets on an interval.
+        """
+        super(Partitioner, self).__init__()
+        self.kz_client = kz_client
+        self.partitioner_path = partitioner_path
+        self.buckets = buckets
+        self.log = log
+        self.got_buckets = got_buckets
+        self.time_boundary = time_boundary
+        ts = TimerService(interval, self.check_partition)
+        ts.setServiceParent(self)
```

Comment:
Instead of inheriting from `MultiService` and doing this, what are you thoughts on inheriting `Service` and keep `TimerService` as an object that gets started and stopped on this service's `startService` and `stopService`? I've done similar thing in [metrics](https://github.com/rackerlabs/otter/blob/master/otter/metrics.py#L317-L353)